src/sentry/sentry_metrics/indexer/ratelimiters.py

```diff
@@ -0,0 +1,169 @@
+        ] + [
+            Quota(prefix_override=None, **args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.releasehealth.per-org")
+        ]
+    else:
+        raise ValueError(use_case_id)
+
+
+@metrics.wraps("sentry_metrics.indexer.construct_quota_requests")
+def _construct_quota_requests(
+    use_case_id: UseCaseKey, keys: KeyCollection
+) -> Tuple[Sequence[OrgId], Sequence[RequestedQuota]]:
+    org_ids = []
+    requests = []
+    quotas = _construct_quotas(use_case_id)
+
+    if quotas:
+        for org_id, strings in keys.mapping.items():
+            org_ids.append(org_id)
+            requests.append(
+                RequestedQuota(
+                    prefix=_build_quota_key(use_case_id, org_id),
+                    requested=len(strings),
+                    quotas=quotas,
+                )
+            )
+
+    return org_ids, requests
+
+
+@dataclasses.dataclass
+class RateLimitState:
+    use_case_id: UseCaseKey
+    requests: Sequence[RequestedQuota]
+    grants: Sequence[GrantedQuota]
+    timestamp: Timestamp
+
+
+@dataclasses.dataclass
+class DroppedString:
+    key_result: KeyResult
+    fetch_type: FetchType
+    fetch_type_ext: FetchTypeExt
+
+
+class WritesLimiter:
+    def __init__(self) -> None:
+        self.rate_limiters: MutableMapping[UseCaseKey, RedisSlidingWindowRateLimiter] = {}
+
+    def _get_rate_limiter(self, use_case_id: UseCaseKey) -> RedisSlidingWindowRateLimiter:
+        if use_case_id not in self.rate_limiters:
+            options = get_ingest_config(use_case_id).writes_limiter_cluster_options
+            self.rate_limiters[use_case_id] = RedisSlidingWindowRateLimiter(**options)
+
+        return self.rate_limiters[use_case_id]
+
+    @metrics.wraps("sentry_metrics.indexer.check_write_limits")
+    def check_write_limits(
+        self, use_case_id: UseCaseKey, keys: KeyCollection
+    ) -> Tuple[RateLimitState, KeyCollection, Sequence[DroppedString]]:
+        """
+        Takes a KeyCollection and applies DB write limits as configured via sentry.options.
+
+        Returns a tuple of:
+
+        1. `RateLimitState` that needs to be passed to `apply_write_limits` in order
+          to commit quotas upon successful DB write.
+
+        2. A key collection containing all unmapped keys that passed through the
+          rate limiter.
+
+        3. All unmapped keys that did not pass through the rate limiter.
+        """
+
+        org_ids, requests = _construct_quota_requests(use_case_id, keys)
+        timestamp, grants = self._get_rate_limiter(use_case_id).check_within_quotas(requests)
+
+        granted_key_collection = dict(keys.mapping)
+        dropped_strings = []
+
+        for org_id, request, grant in zip(org_ids, requests, grants):
+            allowed_strings = granted_key_collection[org_id]
+            if len(allowed_strings) > grant.granted:
+                allowed_strings = set(allowed_strings)
```

Comment:
I know that `allowed_strings` and `granted_key_collection` are mutated through this loop and, in the end they represent what their names say, but that makes this code quite hard to follow as when reading it two variables do not mean what their name says till the end of the loop.

I would suggest this: 
- you start with an empty `allowed_strings` and `dropped_strings`
- take `grant.granted` strings from `granted_key_collection` and put them into `allowed_strings`, the rest goes in `dropped`.

The iterations you skip by mutating `allowed_strings` in place is likely to be irrelevant considering that in the same code path we have done a redis `mget` before getting here.

Reply:
I'm going to roll this back.. I tried to fix the testfailures, and I think this just causes more edgecases to be handled in case there are no quotas configured.

Reply:
I think you would be doing the same amount of clone operations with my suggestion that you are doing now. But again, I don't think it matters as in the same code path (when we got here) we did an mget from redis which is likely to dwarf that impact. 

This is not so critical as it is not in the code path taken by the majority of the lookup operations. THat code path lands on memcache and does not get here.

Reply:
sry @fpacifici, i overlooked this. if you don't mind i will punt on this and see if it becomes a problem again.. i think i painted myself a bit into a corner with the way i am passing data around as well, so i think in order to make this work i would have to change some other signatures as well.